File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def contains(self, r, t):
            if not r:
                return False
            
            if r.val  == t.val:
                return True
            
            return self.contains(r.right, t) or self.contains(r.left, t)
            
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if root.val == p.val or root.val == q.val:
            return root
        elif (self.contains(root.left, p) and self.contains(root.right, q)) or (self.contains(root.right, p) and self.contains(root.left, q)):
            return root
        elif self.contains(root.left, p) and self.contains(root.left, q):
            return self.lowestCommonAncestor(root.left, p, q)
        else:
            return self.lowestCommonAncestor(root.right, p, q)

File: Data Structures & Algorithms/invert-a-binary-tree/submission-6.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <utility>

class Solution {
   public:
    TreeNode* invertTree(TreeNode* root) {
        if (!root) {
            return nullptr;
        }

        std::vector<TreeNode*> s{};
        s.push_back(root);

        while (s.size()) {
            TreeNode* node = s.back();
            s.pop_back();

            std::swap(node->left, node->right);
            if (node->left) {
                s.push_back(node->left);
            }
            if (node->right) {
                s.push_back(node->right);
            }
        }

        return root;
    }
};
